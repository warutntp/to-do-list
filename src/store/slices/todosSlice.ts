import { createSlice, current, PayloadAction } from "@reduxjs/toolkit";
import { TodoItemModels } from "../../types/TodoItemModels";
import {
  loadTodosFromLocalStorage,
  saveTodosToLocalStorage,
} from "../../utils/localStorage";

const initialState: TodoItemModels.TodoItemModel[] =
  loadTodosFromLocalStorage();

const todosSlice = createSlice({
  name: "todos",
  initialState,
  reducers: {
    addTodo: (state, action: PayloadAction<TodoItemModels.TodoItemModel>) => {
      state.push(action.payload);
      let todosData = current(state);
      saveTodosToLocalStorage(todosData);
    },
    removeTodo: (state, action: PayloadAction<string>) => {
      const newState = state.filter(
        (todo: TodoItemModels.TodoItemModel) => todo.slug !== action.payload
      );
      saveTodosToLocalStorage(newState);
      return newState;
    },
    toggleTodo: (state, action: PayloadAction<string>) => {
      const newState = state.map((todo: TodoItemModels.TodoItemModel) =>
        todo.slug === action.payload
          ? { ...todo, completed: !todo.completed }
          : todo
      );
      saveTodosToLocalStorage(newState);
      return newState;
    },
  },
});

export const { addTodo, removeTodo, toggleTodo } = todosSlice.actions;
export default todosSlice.reducer;
