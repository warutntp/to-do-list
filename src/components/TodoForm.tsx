import React, { useState, FormEvent } from "react";
import { nanoid } from "nanoid";
import { addTodo } from "../store/slices/todosSlice";
import { useDispatch } from "react-redux";

const TodoForm: React.FC = () => {
  const [title, setTitle] = useState("");
  const [description, setDescription] = useState("");
  const dispatch = useDispatch();

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (!title.trim()) return;
    const itemData = {
      slug: nanoid(),
      title,
      description,
      completed: false,
    };
    dispatch(addTodo(itemData));
    setTitle("");
    setDescription("");
  };

  return (
    <form onSubmit={handleSubmit} className="mb-3">
      <div className="mb-3">
        <label htmlFor="title" className="form-label">
          Title
        </label>
        <input
          id="title"
          className="form-control w-100"
          type="text"
          value={title}
          onChange={(e) => setTitle(e.target.value)}
          placeholder="Please enter todo title"
          required
        />
      </div>
      <div className="mb-3">
        <label htmlFor="description" className="form-label">
          Description (Optional)
        </label>
        <textarea
          className="form-control"
          id="description"
          rows={3}
          value={description}
          placeholder="Todo description"
          onChange={(e) => setDescription(e.target.value)}
        ></textarea>
      </div>
      <div className="d-flex justify-content-end mb-3">
        <button type="submit" className="btn btn-primary">
          Add Task
        </button>
      </div>
    </form>
  );
};

export default TodoForm;
