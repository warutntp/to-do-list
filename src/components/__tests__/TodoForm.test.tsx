import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import TodoForm from "../TodoForm";
import { addTodo } from "../../store/slices/todosSlice";

const mockStore = configureStore([]);

describe("TodoForm", () => {
  let store: ReturnType<typeof mockStore>;
  let mockDispatch: jest.Mock;

  beforeEach(() => {
    store = mockStore([]);
    mockDispatch = jest.fn();
    store.dispatch = mockDispatch;
  });

  it("should render TodoForm correctly", () => {
    render(
      <Provider store={store}>
        <TodoForm />
      </Provider>
    );

    expect(
      screen.getByPlaceholderText("Please enter todo title")
    ).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText("Enter todo description")
    ).toBeInTheDocument();
    expect(screen.getByText("Add Task")).toBeInTheDocument();
  });

  it("should dispatch addTodo action when form is submitted", () => {
    render(
      <Provider store={store}>
        <TodoForm />
      </Provider>
    );

    fireEvent.change(screen.getByPlaceholderText("Please enter todo title"), {
      target: { value: "New Todo" },
    });
    fireEvent.change(screen.getByPlaceholderText("Enter todo description"), {
      target: { value: "Description" },
    });
    fireEvent.click(screen.getByText("Add Task"));

    expect(mockDispatch).toHaveBeenCalled();
    const action = mockDispatch.mock.calls[0][0];
    expect(action.type).toBe(addTodo.type);
    expect(action.payload).toEqual({
      slug: expect.any(String),
      title: "New Todo",
      description: "Description",
      completed: false,
    });

    expect(screen.getByPlaceholderText("Please enter todo title")).toHaveValue(
      ""
    );
    expect(screen.getByPlaceholderText("Enter todo description")).toHaveValue(
      ""
    );
  });
});
