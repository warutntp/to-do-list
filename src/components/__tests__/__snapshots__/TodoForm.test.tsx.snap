// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoForm should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="mb-3"
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="title"
      >
        Title
      </label>
      <input
        class="form-control placeholder-italic"
        id="title"
        placeholder="Please enter todo title"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="description"
      >
        Description (Optional)
      </label>
      <textarea
        class="form-control placeholder-italic"
        id="description"
        placeholder="Enter todo description"
        rows="3"
      />
    </div>
    <div
      class="mb-3"
    >
      <button
        class="btn btn-primary w-100"
        type="submit"
      >
        Add Task
      </button>
    </div>
  </form>
</DocumentFragment>
`;
